### DISCUSSION TITLE
CAPTCHA or Man-machine verification

### DISCUSSION THREAD

**Discussion by eqfae** [dani-garcia/vaultwarden#2987] [Q&A]
![图片](https://user-images.githubusercontent.com/120385812/207587777-a457672e-b1bc-4251-bb75-ade6afd0a13f.png)
Is there a login verification similar to hCaptcha?

**Comment by BlackDex** [comment]
At the moment not. I know it is supported by Bitwarden, not sure if that also counts for there self-hosted solution though. If that is possible, we may be able to add that, if not, then it will be more difficult.

**Reply by BlackDex** [reply]
Also, we have login rate limiting feature, not totally the same, but it provides some protection.

**Reply by eqfae** [reply]
look forward to

**Reply by sycured** [reply]
My actual setup is hybrid:
- public access only:
    - nginx with rate-limit (https://www.tecmint.com/nginx-rate-limiting/ + http://nginx.org/en/docs/http/ngx_http_limit_req_module.html) with mod_security to protect a little more
- public and private access (lan):
    - rate-limiting at firewall level: https://wiki.nftables.org/wiki-nftables/index.php/Rate_limiting_matchings

Now, I'm looking to implement captcha using [Cloudflare Turnstile](https://developers.cloudflare.com/turnstile/)

**Reply by BlackDex** [reply]
That will probably break all the clients like cli, mobile, desktop.

**Reply by sycured** [reply]
> That will probably break all the clients like cli, mobile, desktop.

On which part?

The WAF, yes, it's a lot of tuning but it works.

The rate-limiting works but I recognize that I need to benchmark on public side with VPN or proxy (zero trust).

The captcha, sorry but passing Cloudflare Turnstile captcha is a lot more easy and quick that hcaptcha. If it breaks, ok, I'll need to implement hcaptcha.

**Reply by BlackDex** [reply]
It's more that if for some reason the client triggers a captcha block they won't show you the captcha screen and block the clients fully.
